pkg/registry/flowcontrol/rest/storage_flowcontrol.go

```diff
@@ -101,70 +100,78 @@ func (p RESTStorageProvider) GroupName() string {
-						return false, nil
-					}
-					return true, nil
-				},
-				hookContext.StopCh)
-			if err != nil {
-				klog.ErrorS(err, "Ensuring suggested configuration failed")
-
-				// We should not attempt creation of mandatory objects if ensuring the suggested
-				// configuration resulted in an error.
-				// This only happens when the stop channel is closed.
-				// We rely on the presence of the "exempt" priority level configuration object in the cluster
-				// to indicate whether we should ensure suggested configuration.
-				return
+	return PostStartHookName, ensureAPFBootstrapConfiguration, nil
+}
+
+func ensureAPFBootstrapConfiguration(hookContext genericapiserver.PostStartHookContext) error {
+	clientset, err := flowcontrolclient.NewForConfig(hookContext.LoopbackClientConfig)
+	if err != nil {
+		return fmt.Errorf("failed to initialize clientset for APF - %w", err)
+	}
+
+	// get a derived context that gets cancelled after 5m or
+	// when the StopCh gets closed, whichever happens first.
+	ctx, cancel := contextFromChannelAndMaxWaitDuration(hookContext.StopCh, 5*time.Minute)
+	defer cancel()
+
+	err = wait.PollImmediateUntilWithContext(
+		ctx,
+		time.Second,
+		func(context.Context) (bool, error) {
+			if err := ensure(clientset); err != nil {
+				klog.ErrorS(err, "APF bootstrap ensurer ran into error, will retry later")
+				return false, nil
 			}
+			return true, nil
+		})
+	if err != nil {
+		return fmt.Errorf("unable to initialize APF bootstrap configuration")
+	}
 
-			const retryCreatingMandatorySettingsInterval = time.Minute
-			_ = wait.PollImmediateUntil(
-				retryCreatingMandatorySettingsInterval,
-				func() (bool, error) {
-					if err := upgrade(
-						flowcontrolClientSet,
-						flowcontrolbootstrap.MandatoryFlowSchemas,
-						// Note: the "exempt" priority-level is supposed to be the last item in the pre-defined
-						// list, so that a crash in the midst of the first kube-apiserver startup does not prevent
-						// the full initial set of objects from being created.
-						flowcontrolbootstrap.MandatoryPriorityLevelConfigurations,
-					); err != nil {
-						klog.Errorf("failed creating mandatory flowcontrol settings: %v", err)
-						return false, nil
-					}
-					return false, nil // always retry
-				},
-				hookContext.StopCh)
-		}()
-		return nil
-	}, nil
+	// we have successfully initialized the bootstrap configuration, now we
+	// spin up a goroutine which reconciles the bootstrap configuration periodically.
+	go func() {
+		err := wait.PollImmediateUntil(
+			time.Minute,
+			func() (bool, error) {
+				if err := ensure(clientset); err != nil {
+					klog.ErrorS(err, "APF bootstrap ensurer ran into error, will retry later")
+				}
+				// always auto update both suggested and mandatory configuration
+				return false, nil
+			}, hookContext.StopCh)
+		if err != nil {
+			klog.ErrorS(err, "APF bootstrap ensurer is exiting")
```

Comment:
this always logs an error on server shutdown... this should be info, at best

Reply:
this confused debugging on https://github.com/kubernetes/kubernetes/issues/103512

Reply:
https://github.com/kubernetes/kubernetes/pull/103550